Fix FTL structs throughout the entire codebase

The-Dumb-Dino opened the pull request:
### Short Description:
This morning I discovered many huge and serious problem with the stripped structs of FTL that breaks parity to the real ones.
To break it down, basically every class that has multiple parents has only one assigned to inherit from, which makes it hard to access certain methods and values.

### Detailed description:
**Background story:**
I have faced a problem like this before where I needed to access a value that was part of one of the many drone structs but was unable to get it because the parser did not assign all the correct parents to the class making it inaccessible, my workaround was to hook a method from the actual parent into the child class which returned what I needed but fixes like this do not always work. 
A few days ago I had another case of this, but this time I didn't have that hacky trick from before as an option, so I decided to do some research and found some problems related to the structs.

**The issue(s):**
Sadly this isn't really just one issue and the individual ones cant be considered small either.

To begin, there are a few issues with parents in structs which I can show with these examples:
![grafik](https://github.com/user-attachments/assets/b4c218c3-a739-4275-a059-e17ef5c34cba)
You can see that there are 2 instances of another struct at the top of the Projectile struct, Collideable and Targetable. When the parsefunc script runs to create the individual header files for each platform it always picks up the members called ``_base`` and makes them the current structs parent to inherit from which is correct, but what isn't correct is that ``_targetable`` is left in there as a member, it actually should be a parent too and I can proof that with ghidra. 
Looking at the decompiled code of a method in the Projectile class we can see that it accesing a value from the ``Targetable`` member as if it was a parent:
![grafik](https://github.com/user-attachments/assets/e500fc34-4e3f-410b-9534-4cf1d76cca64)
Whenever you see something like ``super_*classname*`` in the ghidra decompiler you know that it's reaching for a value provided by a parent.

Now take a look at the struct after it was written into the platform header file with the parsefunc script:
![grafik](https://github.com/user-attachments/assets/7038eebe-0aa7-4f64-a37f-87e3d0b94833)
as you can see, the targetable class is defined as a member and it should not be that way which proofs that the structs we are using in the codebase right now are incorrect. The values inside Targetable can be accesed through that member I believe but it should **not** be like that.

This is not the only case of this issue. Here are a few examples:
![grafik](https://github.com/user-attachments/assets/7b270006-b16c-460e-9bec-1656332edaec)
![grafik](https://github.com/user-attachments/assets/ba04cd11-7bd4-46c2-9acb-6539e99e6eef)
![grafik](https://github.com/user-attachments/assets/5fb94fce-b421-4eac-baac-cbca223993ce)
Every time you find a member whose type is a struct and whose name begins with an underscore, it is actually a parent of the struct it's inside of.

**Solution & the new Issue:**
In my previous commits I updated the parsefunc script and regenerated the headers correctly which now also have their right parents.
This is what the output looks like now:
![grafik](https://github.com/user-attachments/assets/5f252564-d201-4349-b1d9-6661c00d8799) 
![grafik](https://github.com/user-attachments/assets/85afb7fa-4231-430e-8020-1f6b7d4dde2b)
![grafik](https://github.com/user-attachments/assets/5a5f5dbf-6af7-408e-8782-83d139ed48c0)
*Whoooo! Multiple parents :o*

With the structs corrected there comes a new issue. The entire codebase was built with that bug for years accesing the parents through the members which means all the code has to be updated. Luckily this isn't too hard as its mostly just removing the mentions of these prior members. 
![grafik](https://github.com/user-attachments/assets/ec5c3c14-53c6-4c02-a648-e74be3ffe187)
![grafik](https://github.com/user-attachments/assets/b541d06f-88a1-4ef0-974b-00eb543d2160)
![grafik](https://github.com/user-attachments/assets/48695023-8758-4af3-bc48-36a12f3414ea)
This should mean a lot of repetitive, simple work.

The-Dumb-Dino committed:
Adjusted ZHL parser to take care of multi-inheritance

- The parsefunc script can now take assign multiple parents to a struct

The-Dumb-Dino committed:
added another parent keyname to the parser

The-Dumb-Dino committed:
Regenerated Headers with correct parents (BUILD WILL FAIL NOW)

The-Dumb-Dino committed:
Fixed script removing the wrong entries in generated structs

- The previous version only removes the first entry because before it only took care of _base which was always first.
- Added new keyword _selectable which seems to be the last one (finally)
- Altready regenerated the Headers with the new script

The-Dumb-Dino committed:
Revertetd back to original Headers so that debugging them later will be easier

ChronoVortex commented:
> The values inside Targetable can be accesed through that member I believe but it should not be like that.

The severity of this seems a little overplayed. If there are any classes that *don't* have the members of their parents accessible through a field, that is a serious issue. But it sounds like it is all accessible, just formatted incorrectly.

We can certainly scope this in for 2.0.0 if you complete the implementation, but I'm not particularly concerned about it.

The-Dumb-Dino committed:
Got rid of _shipObject & __selectable false members

The-Dumb-Dino commented:
> > The values inside Targetable can be accesed through that member I believe but it should not be like that.
> 
> The severity of this seems a little overplayed. If there are any classes that _don't_ have the members of their parents accessible through a field, that is a serious issue. But it sounds like it is all accessible, just formatted incorrectly.
> 
> We can certainly scope this in for 2.0.0 if you complete the implementation, but I'm not particularly concerned about it.

Sadly this seems to be the issue for some structs.
Several systems, in my example ``EngineSystem`` seems to miss both its parents. Looking at its vtable in ghidra I see many functions from ``ShipSystem`` and ``Selectable`` and yet neither of them exist as a parent or member which is really giving me issues with a branch a created yesterday.

The-Dumb-Dino committed:
Made _targetable a parent of Projectile

The-Dumb-Dino committed:
Made Drone a parent of CrewDrone

- Removed false entry that is actually a parent
- migrated every function using that old entry

The-Dumb-Dino committed:
Made Targetable a parent of SpaceDrone

- Turned false member into actual parent
- Migrated every instance of the old usage
- Added some slight adjustments here and there for better readability

The-Dumb-Dino committed:
Accidentally removed _targetable from ShipManager already

Didn't plan on removing it yet as I didnt take care of it in the code so far.

The-Dumb-Dino committed:
Revert "Accidentally removed _targetable from ShipManager already"

This reverts commit 64671207c90fba4c738dae241e49a90129e9dafb.

The-Dumb-Dino committed:
Revert "Made Targetable a parent of SpaceDrone"

This reverts commit 9e6a0153eae5ce9f44831948ec92ffc8e71ce111.

The-Dumb-Dino committed:
Made _targetable & _collideable parents of SpaceDrone

Ignore previous commits, I had trouble building.
- Succesfully removed all false members in SpaceDrone and turned them into parents.
- Migrates the use of false members.
- Removed false members from lua declaration

The-Dumb-Dino committed:
Reverted Parsefuncs to correct state

turns out my change wasn't needed

The-Dumb-Dino commented:
Alright I fixed every struct except ShipManager, it is giving me a headache and a lot of crashes when starting a run.

Help wanted!
(Don't generate structs with parsefuncs, its set up correctly but I only got so far by manually messing around with structs step after step validating every parent)

The-Dumb-Dino committed:
Merge branch 'master' into fix-ftl-structs-parity-and-inheritance

ranhai613 commented:
Changes in #438 need to be adjusted because the pr has multi inheritance issue.

ranhai613 commented:
And chatgpt recommended me to have those classes as members instead of parents lol.
I assume this is why the author of persefunc didnt make it accept multiple inheritance.

ranhai613 commented:
The reason of incorherent parent class order of ShipManager is because ShipObject is not considered as a polymorphic class in HS code, whereas FTL is.

Currently, ShipObject is defined in HS like this:
```cpp
struct ShipObject
{
  void *vptr;
  int iShipId;
};
```

As you know, ShipObject is onginally a polymorphic class (a class that has virtual functions) since it has vptr, but since none of its virtual functions is exposed to HS, clang considers it is not a polymorphic function.

In Itanium C++ ABI, the layout of multiple inheritance class is determined like this:
1. Primary base (the first base) is the first polymorhic class in declaration order
2. Other bases are layed in declaration order

So, `struct ShipManager : ShipObject, Targetable, Collideable` will be layed down to this order

```
Targetable (the first polymorphic class)
ShipObject (non-polymorphic class)
Collideable
```

And sure enough, I confirmed thats the actuall compiled order by `-fdump-record-layouts` to see layouts

```cpp
*** Dumping AST Record Layout
         0 | struct ShipManager
         0 |   struct Targetable (primary base)
         0 |     (Targetable vtable pointer)
         4 |     int type
         8 |     _Bool hostile
         9 |     _Bool targeted
        12 |   struct ShipObject (base)
        12 |     void * vptr
        16 |     int iShipId
        20 |   struct Collideable (base)
        20 |     (Collideable vtable pointer)
```

The solution is simple: Add virtual functions to ShipObject. Replace `void *vptr` with actually defined vtable, and clang will recognize ShipObject as a polymorphic class then lays it as a primary base of ShipManager.

I confirmed the fix works

```cpp
*** Dumping AST Record Layout
         0 | struct ShipManager
         0 |   struct ShipObject (primary base)
         0 |     (ShipObject vtable pointer)
         4 |     int iShipId
         8 |   struct Targetable (base)
         8 |     (Targetable vtable pointer)
        12 |     int type
        16 |     _Bool hostile
        17 |     _Bool targeted
        20 |   struct Collideable (base)
        20 |     (Collideable vtable pointer)
```

This bug may not only affect ShipManager, as I found seveal `void *vptr;` in Stripped.h. So you will need to replace all of them with actual vtable.
